### PR TITLE
Fix nested computed read

### PR DESCRIPTION
**Pull request by VenelinMartinov**
Fixes an issue with the SDKv2 and PF input extraction. During import we wrongly return values for properties which are fully computed which causes invalid programs to be generated.

~The object check in `CastToTypeObject` was wrong for SDKV2 objects: https://github.com/pulumi/pulumi-terraform-bridge/blob/48fd41bd16e2f7f933bb9390ab7619d52faabb6d/pkg/tfshim/util/types.go#L36~

The `CastToTypeObject` object check is wrong for both SDKv2 and PF. For the SDKV2 only sets and lists can have a `Resource` `.Elem`. For PF it list-nested blocks are represented with a `Resource` `.Elem` in the shim layer, so the check did not catch these. PF details here: https://github.com/pulumi/pulumi-terraform-bridge/pull/2231
This fixes the check and adds regression tests for the broken imports which resulted from 2180.

Test coverage added in https://github.com/pulumi/pulumi-terraform-bridge/pull/2224 for sdkv2 and https://github.com/pulumi/pulumi-terraform-bridge/pull/2225 for pf. 
fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/2180
depends on https://github.com/pulumi/providertest/pull/91
stacked on https://github.com/pulumi/pulumi-terraform-bridge/pull/2225

### CODE REVIEW

**Comment by VenelinMartinov** [comment]
GH rate limits...

File: pkg/tfbridge/schema.go [hunk 1]
```diff
@@ -494,8 +493,6 @@ func (ctx *conversionContext) makeTerraformInput(
 
 		var tfflds shim.SchemaMap
 
-		// We cannot use [shimutil.CastToTypeObject] because we have machinery that constructs invalid
```
Comment (t0yv0) [inline pkg/tfbridge/schema.go:497]: This is exactly the root cause, I find it unfortunate to remove this comment. `elemSchemas` produces

```go
(&schema.Schema{Elem: e}).Shim()
```

This is not a valid schema as it does not have `Type()` defined. This is an artificial schema that you will not find coming from the actual provider, it's a recursion artifact.

File: pkg/tfbridge/schema.go [hunk 2]
```diff
@@ -1722,9 +1719,9 @@ func extractSchemaInputs(
 		return resource.NewArrayProperty(v)
 	case state.IsObject():
 		obj := state.ObjectValue()
-		if tfflds, ok := shimutil.CastToTypeObject(tfs); ok {
+		if tfflds, ok := tfs.Elem().(shim.Resource); ok {
```
Comment (t0yv0) [inline pkg/tfbridge/schema.go:1722]: What happens here is `tfs` does not have a valid Type() because it came from elemSchemas, therefore it fails to pass the CastToTypeObject TypeMap check.

File: pkg/tfbridge/schema_test.go [hunk 3]
```diff
@@ -2735,7 +2787,6 @@ func TestExtractSchemaInputsNestedMaxItemsOne(t *testing.T) {
 				"listObjects": resource.NewProperty([]resource.PropertyValue{
 					resource.NewProperty(resource.PropertyMap{
 						"__defaults": resource.NewProperty([]resource.PropertyValue{}),
-						"field1":     resource.NewProperty(false),
```
Comment (t0yv0) [inline pkg/tfbridge/schema_test.go:2738]: What's happening here? Why this change?

File: pkg/tfbridge/schema.go (hunk 1 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/schema.go:497]: But CastToTypeObject is wrong in the first place for SDKv2 - it checks for .Resource under a Map type which is illegal in SDKV2.

I don't understand why we'd want to use it here anyway - `tfs.Elem().(shim.Resource)` is the correct way to check for an object type

File: pkg/tfbridge/schema_test.go (hunk 3 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/schema_test.go:2738]: field1 is a boolean and we have logic which gets rid of default values for basic types when constructing inputs from outputs:
https://github.com/pulumi/pulumi-terraform-bridge/blob/18db7c57211af0eb7dfa26a70b145edc4d33a88f/pkg/tfbridge/schema.go#L1657

called from 

https://github.com/pulumi/pulumi-terraform-bridge/blob/18db7c57211af0eb7dfa26a70b145edc4d33a88f/pkg/tfbridge/schema.go#L1773

Seems like it was the original intention here.

File: pkg/tfbridge/schema.go (hunk 1 above)
Comment (t0yv0) [reply pkg/tfbridge/schema.go:497]: I'm not going to block on this change but the Elem() representation is continuously confusing.  I don't think "object type" is defined enough for me. 

How about https://github.com/pulumi/pulumi-terraform-bridge/pull/2187 to try to clarify some of the issues here?

File: pkg/tfbridge/schema_test.go (hunk 3 above)
Comment (t0yv0) [reply pkg/tfbridge/schema_test.go:2738]: Gotcha. So this got fixed because the fix now propagates schema properly.

**Review by t0yv0 — APPROVED** [review]

File: pkg/tfshim/util/types.go [hunk 4]
```diff
@@ -31,6 +31,7 @@ func IsOfTypeMap(tfs shim.Schema) bool {
 
 // CastToTypeObject performs a checked cast from shim.Schema to a TF object (a collection
 // of fields).
+// Note that this is only valid for PF resources - in the TF SDK objects are not represented as maps
```
Comment (t0yv0) [inline pkg/tfshim/util/types.go:34]: Confusing... Not going to block on this but here's my attempt to update docu https://github.com/pulumi/pulumi-terraform-bridge/pull/2187 on Elem instead around this. I looked at call sites of this function there's 4, mostly look reasonable-ish.

File: pkg/tfbridge/schema_test.go [hunk 5]
```diff
@@ -2172,6 +2172,58 @@ func TestRefreshExtractInputsFromOutputsMaxItemsOne(t *testing.T) {
...
+				"key256": (&schema.Schema{Type: shim.TypeString, Computed: true}).Shim(),
+			}),
+		}
+	}
+
+	ruleSetPs := func() map[string]*SchemaInfo {
+		return map[string]*SchemaInfo{
+			"attachedDisks": {
+				Fields: map[string]*SchemaInfo{
+					"name":   {},
```
Comment (iwahbe) [inline pkg/tfbridge/schema_test.go:2211]: In `ruleSetProps` `name` is at the path `attachedDisks[0].name`. `ruleSetSchema` also defines `name` with `attachedDisks: list({"name", "key256"})`.

`ruleSetPs` defines name at `["attachedDisks"].Fields["name"]` (without the list index). I think it should be `["attachedDisks"].Elem.Fields["name"]`.

If you want to make sure that a schema override propagated, I would use a `Name:` override for `"name"`: `"name": {Name: "renamed"}`.

File: pkg/tfbridge/schema.go (hunk 1 above)
Comment (iwahbe) [reply pkg/tfbridge/schema.go:497]: My understanding was that `tfs.Elem().(shim.Resource)` was insufficient to identify an object type: `tfs.Type() == TypeList && tfs.Elem().(shim.Resource)` implied a `List[Object]` where `tfs.Type() == TypeMap && tfs.Elem().(shim.Resource)` implied an `Object` type.

File: pkg/tfshim/util/types.go (hunk 4 above)
Comment (iwahbe) [inline pkg/tfshim/util/types.go:34]:
```suggestion
// CastToTypeObject is only valid for PF resources - in the TF SDK objects are not represented as maps.
```

File: pkg/tfbridge/schema.go (hunk 1 above)
Comment (t0yv0) [reply pkg/tfbridge/schema.go:497]: When recurring over a List[Object] type which is indeed `tfs.Type() == TypeList && tfs.Elem().(shim.Resource)` the code will call elemSchemas to compute a shim.Schema for the Object, which will result in a shim.Schema value that's not recognized by CastToTypeObject. One can argue that it should, in fact, be recognized, and then we could bring that back here and work this out that way. That's a bit more changes though.

**Comment by VenelinMartinov** [comment]
After adding the PF tests it looks like this affects PF too. I'll add that to the issue and PR description.

File: pkg/tfshim/util/types.go (hunk 4 above)
Comment (VenelinMartinov) [reply pkg/tfshim/util/types.go]: https://github.com/pulumi/pulumi-terraform-bridge/pull/2231#discussion_r1687740016

Looks like `CastToTypeObject` does not catch all objects in PF either. I'll raise a PR for this but I wonder how this affects other uses.
